### PR TITLE
[TRTLLM-16104][test] Add MX weight manifests and widen the ModelExpress qualification probe

### PR DESCRIPTION
**Pull request by moraxu**
## Summary

Pre-merge tier of the [MX Model Family Testing Proposal](https://docs.google.com/document/d/14SZmmFcoakgIx2OC4dt8pWcHU14PDTN9KlAKqLoZ15s/edit?tab=t.1m276imwr1p4) (Lever 1). Today CI never checks that ModelExpress (MX) P2P-loaded weights equal HF-loaded weights on real models: the unit harness proves exact-value equality on tiny CPU fixtures, and the E2E test compares 8 greedy tokens on 2 prompts of TinyLlama. Staged-hook ordering bugs are deterministic weight corruption, so this PR adds a byte-exact detector where it is cheapest and widens the behavioral probe.

- **Weight manifest** (`tensorrt_llm/_torch/weight_sharing/weight_manifest.py`): SHA-256 per registered parameter/buffer over canonical bytes (`t.detach().reshape(-1).contiguous().cpu().view(torch.uint8)`), plus dtype/shape/stride/storage_offset metadata, storage-alias partitions, skipped-tensor records, a whole-manifest digest, and `manifest_format_version`. Strictly stronger than `assert_close(rtol=0, atol=0, equal_nan=True)` (distinguishes signed zeros and NaN payloads). Inert unless `MX_WEIGHT_MANIFEST_DIR` + `MX_WEIGHT_MANIFEST_ROLE` are set; when active, every problem raises.
- **Two capture points**: the `final` manifest at the single return of `ModelLoader.load` (after all post-load hooks and MoE load-balancer finalization, before warmup — the one path every role shares; cost recorded as `weight_manifest_seconds`), and the `transfer` manifest inside `MXCheckpointLoader` at receiver P2P success (now CUDA-synchronized first) and donor publish (outside the best-effort publish `try`).
- **E2E harness**: helpers move to `mx_harness.py` + stdlib-only `mx_evidence.py` (`model_express/` becomes a package). `test_mx_donor_receiver` now enforces the transfer tier (donor@publish == receiver@receive, parameters) and the final tier (baseline/donor/receiver pairwise byte-identical incl. buffers, skipped sets, alias partitions), with only a per-row documented `final_manifest_exempt_patterns` escape hatch (never a tolerance). The behavioral probe grows from 2×8 to **8 prompts × 32 greedy tokens** (`max_seq_len` 64→128, `max_num_tokens` 64→256). Payloads, logs, manifests, and `timing.json` are archived under `--output-dir` (always set in CI), including on failure.
- **Unit tests**: `test_weight_manifest.py` (contract + corruption-injection trio + non-contiguous/alias/meta/version/round-trip/env-gating), hook tests in `test_model_loader_mx.py` and `test_mx_checkpoint_loader.py`, `test_mx_evidence.py`, and a byte-level tightening of `tests/unittest/utils/post_transform_qualification.py`.

Design decisions worth a look: the transfer tier compares **parameters only** (`TRANSFER_TIER_KINDS`) because the receiver's `cache_derived_state()` runs after the P2P boundary, so derived buffers are enforced at the final tier; flip the constant if the first H100 run shows buffers are also identical at the boundary.

## Test Coverage

Unit (CPU): `tests/unittest/_torch/weight_sharing/test_weight_manifest.py`, `tests/unittest/_torch/weight_sharing/test_mx_evidence.py`, additions in `tests/unittest/_torch/executor/test_model_loader_mx.py` and `tests/unittest/_torch/models/checkpoints/mx/test_mx_checkpoint_loader.py`; existing lifecycle tests now also assert byte equality.

E2E: `model_express/test_model_express.py::test_mx_donor_receiver[*]` on `DGX_H100-2_GPUs-PyTorch-ModelExpress-1` (TP1) and `DGX_H100-4_GPUs-PyTorch-ModelExpress-OnDemand-1` (TP2).

Measured on the H100 stage (to fill in): per role/rank `manifest_final_seconds`, `manifest_transfer_seconds`, `bytes_hashed`; `load_seconds`/`generate_seconds` before vs after the widening; per-test wall time before (`.test_durations`: llama tp1 149 s, tp2 157 s) vs after.

## PR Checklist

- [x] Follows `CODING_GUIDELINES.md`; NVIDIA header on new files; `git commit -s`
- [x] Unit tests added; E2E test updated
- [x] Docs updated (`docs/source/features/model-express.md`)
- [x] JIRA number to be added to the title
- [x] Rebase onto main after #18558 (Mistral MX) merges; only `_MX_CASES` tail and the docs paragraph can conflict

Stacked follow-up: post-merge accuracy canaries (Lever 2/3 of the design doc).

## Dev Engineer Review

- Adds environment-gated, byte-exact SHA-256 manifests for ModelExpress parameters and buffers.
- Records metadata, aliases, skipped tensors, format versions, and manifest digests.
- Captures final manifests after loading and transfer manifests at MX P2P boundaries.
- Adds public APIs, atomic writes, comparison logic, timing metrics, and failure cleanup.
- Review focus: manifest lifecycle, transfer compatibility, reload behavior, synchronization, and performance.
- DGX H100 CI runs remain failing and require investigation.

## QA Engineer Review

- Adds unit tests for hashing, serialization, aliases, skipped tensors, corruption, comparison, gating, CUDA synchronization, loader metadata, transfer failures, fallback behavior, and evidence parsing.
- Expands the ModelExpress probe to eight prompts and 32 generated tokens.
- Validates transfer evidence, manifest equality, final byte identity, timing output, and failure artifact retention.
- Adds AST-based import validation for `mx_evidence.py`, including third-party, nested, local, and relative imports.
- The ModelExpress integration test remains listed in `tests/integration/test_lists/test-db/l0_model_express.yml`. Unit tests are not separately listed.
- Coverage verdict: sufficient, with follow-up required for failing DGX H100 CI runs.

## Per-File QA Perspective

- `docs/source/features/model-express.md`: Documents manifest requirements, timing, exemptions, compatibility, and archived artifacts.
- `tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py`: Adds donor and successful receiver manifest writes. Verify role validation, source identity checks, and fallback behavior.
- `tensorrt_llm/_torch/pyexecutor/model_loader.py`: Writes final manifests during `load()` and records `weight_manifest_seconds`. Verify gating and unchanged `reload()` behavior.
- `tensorrt_llm/_torch/weight_sharing/__init__.py`: Exposes the manifest API and types. Verify downstream import compatibility.
- `tensorrt_llm/_torch/weight_sharing/weight_manifest.py`: Adds hashing, serialization, atomic writes, loading, comparison, alias handling, skip rules, and environment-controlled generation. Verify metadata validation and failure cleanup.
- `tests/integration/defs/model_express/__init__.py`: Adds the shared ModelExpress package initializer. It is not a test-list entry.
- `tests/integration/defs/model_express/mx_e2e_worker.py`: Expands probe inputs and propagates manifest settings. Verify deterministic prompts, token limits, and worker environments. It is exercised by the listed ModelExpress integration test.
- `tests/integration/defs/model_express/mx_evidence.py`: Parses and validates RDMA evidence. Verify malformed, duplicate, missing, and failed transfers. It is covered by unit tests and is not separately listed.
- `tests/integration/defs/model_express/mx_harness.py`: Adds lifecycle, manifest, evidence, timing, and artifact helpers. Verify cleanup and failure artifact retention. It is exercised by the listed ModelExpress integration test.
- `tests/integration/defs/model_express/test_model_express.py`: Validates baseline, donor, and receiver behavior through the shared harness. It remains listed in `test-db`.
- `tests/unittest/_torch/executor/test_model_loader_mx.py`: Covers final manifests, receiver metadata, disabled output, and rank mapping. It is not separately listed.
- `tests/unittest/_torch/models/checkpoints/mx/test_mx_checkpoint_loader.py`: Covers transfer manifests, publication failures, fallback paths, role validation, and duplicate prevention. It is not separately listed.
- `tests/unittest/_torch/weight_sharing/test_mx_evidence.py`: Covers evidence parsing, rank validation, failure markers, duplicates, JSON records, and AST import validation. It is not separately listed.
- `tests/unittest/_torch/weight_sharing/test_weight_manifest.py`: Covers hashing, aliases, serialization, comparison, validation, environment gating, skipped-record differences, and CUDA synchronization. It is not separately listed.
- `tests/unittest/utils/post_transform_qualification.py`: Adds canonical byte comparison after value checks. Verify that byte-different representations fail as intended. It is not separately listed.

### CODE REVIEW

**Comment by moraxu** [comment]
@coderabbit fullreview

**Comment by moraxu** [comment]
@coderabbit fullreview

**Comment by moraxu** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-2_GPUs-PyTorch-ModelExpress-1,DGX_H100-4_GPUs-PyTorch-ModelExpress-OnDemand-1"

**Comment by moraxu** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-2_GPUs-PyTorch-ModelExpress-1,DGX_H100-4_GPUs-PyTorch-ModelExpress-OnDemand-1"

File: tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py [hunk 1]
```diff
@@ -573,6 +597,10 @@ def load_weights(self, checkpoint_dir: str, mapping: Mapping, **kwargs) -> dict[
             return fallback_weights
 
         self._p2p_succeeded = True
+        # P2P writes and any upstream dtype casts must be globally visible
+        # before the bytes are fingerprinted or finalized by ModelLoader.
+        _synchronize_cuda_for_mx_publish()
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py:602]: The new receiver-side `_synchronize_cuda_for_mx_publish()` is outside the manifest environment gate, so every successful production MX receiver load pays a device-wide synchronization even when `MX_WEIGHT_MANIFEST_DIR` is unset. That contradicts the no-op contract, and enabled qualification runs synchronize twice because `build_weight_manifest()` already synchronizes all CUDA devices before hashing. 

We could remove or gate the receiver-only sync and rely on the manifest builder while capture is active.

File: tensorrt_llm/_torch/weight_sharing/weight_manifest.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+    for kind in kinds:
+        if kind not in WEIGHT_MANIFEST_KINDS:
+            raise ValueError(
+                f"Unknown tensor kind {kind!r}; expected one of {WEIGHT_MANIFEST_KINDS}"
+            )
+    patterns = tuple(exempt_patterns)
+
+    expected_entries = expected.entries_by_fqn(kinds)
+    actual_entries = actual.entries_by_fqn(kinds)
+    expected_skipped = {(item.fqn, item.reason) for item in expected.skipped if item.kind in kinds}
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/weight_sharing/weight_manifest.py:558]: `SkippedTensor` records `kind`, `dtype`, and `shape`, but the comparator drops all three and compares only (`fqn`, `reason`). Two roles can therefore skip the same FQN for the same reason while disagreeing on whether it is a parameter or buffer, its dtype, or its shape, and the qualification still passes. 

Please compare the complete skipped records, or at least FQN/kind/reason/dtype/shape, and add a same-reason/different-metadata regression test.

File: tests/integration/defs/model_express/test_model_express.py [hunk 3]
```diff
@@ -159,437 +108,102 @@ class MxE2ECase:
...
+        assert tokens["donor"] == tokens["baseline"], "MX donor tokens differ from the HF baseline"
+        assert tokens["receiver"] == tokens["baseline"], (
+            "MX receiver tokens differ from the HF baseline"
+        )
+        assert_transfer_evidence(
+            case, layout.log("receiver"), layout.transfer_logs("receiver"), payloads["receiver"]
+        )
+        manifests = collect_weight_manifests(layout.manifest_dir, case)
+        assert_weight_manifests(case, manifests)
+        report_timings(case, payloads, manifests, layout)
```
Comment (chienchunhung) [inline tests/integration/defs/model_express/test_model_express.py:207]: nit: The docs promise that `timing.json` is archived even when the qualification fails, but `report_timings()` runs only after every token, transfer-evidence, and manifest assertion succeeds. A failure therefore archives no timing file. Let's try to emit best-effort partial timing data from the cleanup path.

**Review by chienchunhung — APPROVED** [review]
Overall LGTM. Thanks!

File: tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py (hunk 1 above)
Comment (BowenFu) [reply tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py:602]: +1

File: tensorrt_llm/_torch/weight_sharing/weight_manifest.py (hunk 2 above)
Comment (BowenFu) [reply tensorrt_llm/_torch/weight_sharing/weight_manifest.py:558]: +1

**Review by yufeiwu-nv — APPROVED** [review]

File: tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py (hunk 1 above)
Comment (moraxu) [reply tensorrt_llm/_torch/models/checkpoints/mx/checkpoint_loader.py]: Thanks, addressed in https://github.com/NVIDIA/TensorRT-LLM/pull/18560/commits/e1a777e797dc642bb081c25dbb42abc1d29078fa

File: tensorrt_llm/_torch/weight_sharing/weight_manifest.py (hunk 2 above)
Comment (moraxu) [reply tensorrt_llm/_torch/weight_sharing/weight_manifest.py]: Thanks, addressed in https://github.com/NVIDIA/TensorRT-LLM/pull/18560/commits/e1a777e797dc642bb081c25dbb42abc1d29078fa

File: tests/integration/defs/model_express/test_model_express.py (hunk 3 above)
Comment (moraxu) [reply tests/integration/defs/model_express/test_model_express.py]: Thanks, addressed in https://github.com/NVIDIA/TensorRT-LLM/pull/18560/commits/e1a777e797dc642bb081c25dbb42abc1d29078fa

**Comment by moraxu** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-2_GPUs-PyTorch-ModelExpress-1,DGX_H100-4_GPUs-PyTorch-ModelExpress-OnDemand-1"

**Review by BowenFu — APPROVED** [review]

**Comment by moraxu** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-2_GPUs-PyTorch-ModelExpress-1,DGX_H100-4_GPUs-PyTorch-ModelExpress-OnDemand-1"

**Comment by moraxu** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-2_GPUs-PyTorch-ModelExpress-1,DGX_H100-4_GPUs-PyTorch-ModelExpress-OnDemand-1"

**Review by arysef — APPROVED** [review]

**Comment by brnguyen2** [comment]
Triaged the blossom-ci failure (L0_MergeRequest_PR #59978, 18 failed):

**Caused by this PR** — `test_timing_report_and_archive_survive_a_failed_run` (`tests/unittest/_torch/weight_sharing/test_mx_harness_artifacts.py`) errors at setup on CPU-Generic-x86-1 and CPU-Generic-arm-1 with `cannot use capsys and capfd at the same time`. Root cause: `tests/unittest/conftest.py` has an autouse fixture `cuda_error_early_quit(capfd)`, so every test in that tree already holds `capfd` and pytest rejects a test that also requests `capsys`. Fix: request `capfd` instead (same stdout capture, fd-level).

I can't push to this branch, so the one-commit fix is staged as a mirror draft: #19105 — cherry-pick `9de5c30b593ab63c8a369d42971e7d7d1d45d152` from `brnguyen2:mirror-18560` and it should go green on those stages.

**Not caused by this PR** (known main flakes, per the infra stability report, last 30 days):
- `TestQwen3LoRA::test_qwen3_bf16_lora_overlap` (B300-PyTorch-1): 2.8% fail rate, 509 MRs, 23 distinct users.
- `test_kimi_k3_situ_moe.py` (DGX_B200-PyTorch-9): 3.3% fail rate, 365 MRs, 22 distinct users.

**Comment by brnguyen2** [comment]
Pushed the fix directly to this branch (9de5c30): the new `test_timing_report_and_archive_survive_a_failed_run` used `capsys`, which collides with the repo-wide autouse `capfd` fixture (`cannot use capsys and capfd at the same time`) — switched it to `capfd`. Rekicking CI to pick it up.

**Review by BowenFu — APPROVED** [review]

**Review by aswinvisva — APPROVED** [review]

**Review by tburt-nv — APPROVED** [review]

File: tensorrt_llm/_torch/weight_sharing/weight_manifest.py [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+
+
+def write_weight_manifest(manifest: WeightManifest, path: Path) -> None:
+    """Atomically write `manifest` to `path`; refuse to overwrite an existing file."""
+    path = Path(path)
+    if path.exists():
+        raise FileExistsError(f"Weight manifest already exists: {path}")
+    path.parent.mkdir(parents=True, exist_ok=True)
+    temporary = path.with_name(path.name + ".tmp")
+    temporary.write_text(serialize_weight_manifest(manifest), encoding="utf-8")
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/weight_sharing/weight_manifest.py]: should we add `try... except...` to remove the possible remaining  `.tmp` file, if the writing process is failed.

**Review by Wanli-Jiang — APPROVED** [review]
LGTM for torch/model part.

File: tensorrt_llm/_torch/weight_sharing/weight_manifest.py (hunk 4 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/weight_sharing/weight_manifest.py]: Good point — done in 35bb7050f7: the write/rename is wrapped so a failure unlinks the `.tmp` (`missing_ok=True`) and re-raises. Added `test_failed_write_leaves_no_tmp_file`, which injects a rename failure after the `.tmp` exists and asserts nothing survives.

**Review by allisonlim-nv — APPROVED** [review]
